libs/wingc/src/type_check.rs

```diff
@@ -1299,7 +1419,19 @@ impl Types {
 
 		let existing_type = self.inferences.get_mut(id).expect("Inference id out of bounds");
 
-		assert!(existing_type.is_none(), "Inference id already has a type");
+		if let Some(existing_type) = existing_type {
+			// if the types are the same, ok, otherwise error
+			if std::ptr::eq(&**existing_type, &*new_type) {
```

Comment:
str::ptr::eq only checks if the values to the same internal instance of `Type`, not necessarily if the they're logically the same in Wing's type system. For example its possible to have two `Type` instances of `Type::Array(Type::Number)` with different pointers. Is that the intention here? (otherwise it might make sense to use `existing_type.is_same_type_as(new_type)`)

Reply:
It was the intention, in theory the only time this should legally happen is when they are the same instance. I can explore this with an additional test case though and see if it should change.